lib/stateTransition/AbstractStateTransition.js

```diff
@@ -49,6 +77,154 @@ class AbstractStateTransition {
   hash() {
     return hash(this.serialize()).toString('hex');
   }
+
+  /**
+   *  Returns signature
+   *
+   * @return {string|null}
+   */
+  getSignature() {
+    return this.signature;
+  }
+
+  /**
+   * Returns public key id
+   *
+   * @returns {number|null}
+   */
+  getSignaturePublicKeyId() {
+    return this.signaturePublicKeyId;
+  }
+
+  /**
+   * Sign data with privateKey
+   *
+   * @param {IdentityPublicKey} identityPublicKey
+   * @param {string} privateKey
+   * @return {AbstractStateTransition}
+   */
+  sign(identityPublicKey, privateKey) {
+    const data = this.serialize({ skipSignature: true });
+    const privateKeyInHex = Buffer
+      .from(privateKey, 'base64')
+      .toString('hex');
+
+    switch (identityPublicKey.getType()) {
+      case IdentityPublicKey.TYPES.ECDSA_SECP256K1:
+        this.signature = signData(data, privateKeyInHex).toString('base64');
+        break;
+      default:
+        throw new InvalidSignatureTypeError(identityPublicKey.getType());
+    }
+
+    const extractedPublicKey = new PrivateKey(privateKeyInHex)
+      .toPublicKey()
+      .toBuffer()
+      .toString('base64');
+
+    if (extractedPublicKey !== identityPublicKey.getData()) {
+      throw new InvalidSignaturePublicKeyError(identityPublicKey.getData());
+    }
+
+    this.signaturePublicKeyId = identityPublicKey.getId();
+
+    return this;
+  }
+
+  /**
+   *
+   * @param {string} privateKey
+   * @return {AbstractStateTransition}
+   */
+  signByPrivateKey(privateKey) {
+    const data = this.serialize({ skipSignature: true });
+    const privateKeyInHex = Buffer
+      .from(privateKey, 'base64')
+      .toString('hex');
+
+    this.signature = signData(data, privateKeyInHex).toString('base64');
+
+    return this;
+  }
+
+  /**
+   * Verify signature
+   *
+   * @param {IdentityPublicKey} publicKey
+   * @return {boolean}
+   */
+  verifySignature(publicKey) {
+    const signature = this.getSignature();
+    if (!signature) {
+      return false;
+    }
+
+    const data = this.serialize({ skipSignature: true });
+
+    try {
+      const publicKeyBuffer = Buffer.from(publicKey.getData(), 'base64');
+      // eslint-disable-next-line no-underscore-dangle
+      const publicKeyHash = new PublicKey(publicKeyBuffer)._getID();
+
+      return verifyDataSignature(
+        data,
+        Buffer.from(signature, 'base64'),
+        publicKeyHash,
+      );
+    } catch (e) {
+      return false;
+    }
+  }
+
+  /**
+   * Verify signature with public key
+   * @param {string} publicKey
+   * @returns {boolean}
+   */
+  verifySignatureByPublicKey(publicKey) {
+    const signature = this.getSignature();
+    if (!signature) {
+      return false;
+    }
+    const data = this.serialize({ skipSignature: true });
+
+    try {
+      const publicKeyBuffer = Buffer.from(publicKey, 'base64');
```

Comment:
The same as for `sign`. Let's accept what PublicKey constructor accepts without decoding.